### PR TITLE
Collect timing information about WASM execution in reducers

### PR DESCRIPTION
**Pull request by gefjon**
# Description of Changes

This commit augments our logging of reducer execution time by logging the time spent specifically executing WASM code, in addition to the overall elapsed time.

This is accomplished by adding two time-related fields to `WasmInstanceEnv`, which we interact with in four contexts:
- before calling into a reducer.
- after returning from a reducer.
- after entering a host function called by a reducer.
- before returning from a host function called by a reducer.

The timings are probably not terribly precise,
as we necessarily capture a small amount of overhead during calls and returns, but they're good enough for our purposes,
i.e. knowing whether slow reducers are slow
because they spend a lot of time in the datastore, or because they spend a lot of time doing WASM computations.

EDIT: Converting to draft, as `perf` profiling seems to show this has non-negligible overhead when running BitCraft - slightly over 1% of time spent in a certain reducer we're concerned about. We'll need to figure out whether to feature-gate, whether the info is valuable enough to make the overhead worthwhile, or whether we can collect the same info in a more efficient way.

# API and ABI

 - [ ] This is a breaking change to the module ABI
 - [ ] This is a breaking change to the module API
 - [ ] This is a breaking change to the ClientAPI
 - [ ] This is a breaking change to the SDK API

*If the API is breaking, please state below what will break*

### CODE REVIEW

**Comment by gefjon** [comment]
benchmarks please

File: crates/core/src/host/wasmer/wasm_instance_env.rs [hunk 1]
```diff
@@ -35,9 +47,79 @@ fn mem_err(err: MemoryAccessError) -> RuntimeError {
...
+/// This version mutably borrows `caller` once before `body`, and again after,
+/// but does not hold the borrow during `body`.
+macro_rules! with_timing_paused {
+    ($caller:ident -> { $($body:tt)* }) => {{
+        $caller.data_mut().timing_after_exit_wasm();
+
+        // Improvised try block: errors bubbled with `?` are captured in `res`,
+        // and the `timing_before_enter_wasm` runs regardless.
+        //
+        // Clippy will warn about this if `body` doesn't include a `?` or `return`,
```
Comment (kulakowski) [inline crates/core/src/host/wasmer/wasm_instance_env.rs:74]: I'm no expert at the interactions of clippy and macros but this seems fine to me.

**Review by kulakowski — APPROVED** [review]

File: crates/core/src/host/wasmer/wasmer_module.rs [hunk 2]
```diff
@@ -377,10 +389,23 @@ impl WasmerInstance {
...
+
+        // Close the `execution_duration` span and collect the overall execution time.
+        // Unlike `wasm_duration`, this includes time spent in host functions,
+        // e.g. evaluating queries.
+        let duration = start.elapsed();
+
+        // Clear buffers and timing spans to avoid leaking data across reducer runs.
         self.env.as_mut(store).buffers.clear();
+        self.env.as_mut(store).iters.clear();
+        self.env.as_mut(store).timing_spans.clear();
```
Comment (kulakowski) [inline crates/core/src/host/wasmer/wasmer_module.rs:404]: I'm now wondering why you are doing this here but not other places that were calling `self.env.as_mut(store).buffers.clear()`, specifically in `call_describer`. I guess that we just don't care about the timing in that path right now?

Also, why `self.env.as_mut(store).iters.clear()` is now being called here now.

IMO the lack of encapsulation between wasm_instance_env and instance_env makes this challenging to understand.

Comment (gefjon) [reply crates/core/src/host/wasmer/wasmer_module.rs:404]: Ah, I didn't mean to commit this change. At some point we should investigate what state we do and don't persist between calls, as I think we're probably leaking some of it that we don't want to, but I think that's out of scope.

**Comment by gefjon** [comment]
Converting to draft, as perf profiling seems to show this has non-negligible overhead when running BitCraft - slightly over 1% of time spent in a certain reducer we're concerned about. We'll need to figure out whether to feature-gate, whether the info is valuable enough to make the overhead worthwhile, or whether we can collect the same info in a more efficient way.

**Pull request closed without merging**